### ISSUE TITLE
Get-PnPProvisioningTemplate : The file is not checked out.  You must first check out this document before making changes.

### ISSUE THREAD

**Issue by Pieter-Veenstra** [SharePoint/PnP-Sites-Core#863]
#### Category
[X ] Bug
[ ] Enhancement

#### Environment
[X] Office 365 / SharePoint Online
[ ] SharePoint 2016
[ ] SharePoint 2013

#### Expected or Desired Behavior
Exporting a site should not require pages to be checked out. This used to work ok in the past.

#### Observed Behavior
Get-PnPProvisioningTemplate : The file is not checked out.  You must first check out this document before making changes.
At C:\PVSource\updateAll.ps1:12 char:1
+ Get-PnPProvisioningTemplate -Out C:\PVSource\templ ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (:) [Get-PnPProvisioningTemplate], ServerException
    + FullyQualifiedErrorId : EXCEPTION,SharePointPnP.PowerShell.Commands.Provisioning.GetProvisioningTemplate

#### Steps to Reproduce
Create publishing site add and run Get-PnPProvisioningTemplate

**Comment by camiloborges** [comment]
Hello team.

Pieter: This seems to be a bug and it happens for any page that requires check out.  One workaround is to check out pages before getting template(you would need to write a script to go through pages, check them out, and then another script to undo checkout upon finishing the templating). I've done this for a specific scenario and it works for me., but I am not sure about all the ramifications of that. 

Core Team.
This exception happens on PageExtensions.cs, line 424.  code reads
````c#
                    if (exportMode != WebParts.WebPartExportMode.All)
                    {
   422:                     wp.WebPart.ExportMode = WebParts.WebPartExportMode.All;
   423:                     wp.SaveWebPartChanges();
   424:                     web.Context.ExecuteQueryRetry();
                        changed = true;
                    }
```

I am writing an object handler and I am handling the check out status at that layer and then the method getwebpartxml always assumes it is ok to change a web part. Essentially files that require check out need to be checked out before changing web part export mode. I believe the 'templating' process shouldn't change anything in the system, so I am always checking pages out, extract, then I undo the checkout. 
I hope to have a PR with a fix for this before the end of the month, but depends on how much time I will be able to commit to it:D

Cheers

**Comment by saurabhjatara** [comment]
I Tried  "-ExcludeHandlers Pages" this even didn't work for me ..

**Comment by Pieter-Veenstra** [comment]
The easy workaround here is to check out the file within the script:

> Set-PnPFileCheckedOut -Url /sites/subsite/subweb/Pages/default.aspx
> Get-PnPProvisioningTemplate -Out $templateFile   -Web subweb  -ExcludeHandlers Publishing -Force

As I'm collecting the template from a template site anyway it doesn't matter to leave it checked out. I could of could also use a Set-PnPFileCheckIn  to get the page back into it's original state.

**Comment by saurabhjatara** [comment]
Thanks @Pieter-Veenstra , but i m trying to get the template for site collection so what are the files i have to checkout ?

**Comment by Pieter-Veenstra** [comment]
Hi @saurabhjatara , If you run your powerShell in debug mode using Set-PnPTraceLog(https://veenstra.me.uk/2016/07/29/office-365-sharepoint-how-to-debug-pnp-provisioning-engine/)

then you will find out which file is causing the problem (in my case only the welcome page of the site).  Then you need to check out the file mentioned in the debug information.

**Comment by saurabhjatara** [comment]
Thanks @Pieter-Veenstra , It helped me a lot.

**Comment by erwinvanhunen** [comment]
Hi, 

I fixed this with PR #950 . If the issue still stands do not hesitate to open this issue again.
